Clang format as a build target

jroweboy opened the pull request:
Please review it commit by commit. This also depends on https://github.com/yuzu-emu/ext-windows-bin/pull/7 for letting windows download a clang-format binary that they can use directly.

## What is all here

* Generates an optional build target when cmake can find the right version of clang-format. For MSVC, you can right click the clang-format project and have it format all source files. For Makefiles, you can run `make clang-format`. I don't know what XCode does as I don't ever use it.

* Each platform has a custom command that is run to gather all the files when the clang-format is run and not when the project is generated. ~~Its *probably* fine to gather a list of all source files using `file(GLOB)` at configuration time, but I was worried that I might miss a case where someone adds a file without regenerating the build scripts. I won't fight it if someone thinks this should be simplified.~~ Eh, I was working on this, but I am running into command line length limits and I don't care to fix that.

* Bumps up the version to clang-format 6.0 which is the version installed when installing the MSVC plugin.

* For a bonus change, I added conditional options to automatically turn on `USE_BUNDLED_*` when MSVC is the toolset and the `ENABLE_*` option is ON. It always seemed silly to me that we make you click those check boxes ... After this is merged, I'll remove those steps from the wiki.

## If you would like to test this:

I've updated the instructions on the contributing guide with how to run clang format. Please try it out and let me know if anything is unclear.

MerryMage commented:
* Is `CMakeDependentOption` not a built-in module?
* Specify that clang-format **6.0** is used in the contrib guide since different versions have different behaviours.
* Fairly sure this doesn't work on macOS (currently away from a mac for two days). If one wants to merge this before I check for contrib reasons, I don't mind, I'll check when I'm able.
* Good timing. macbrew removed the `clang-format@3.8` package yesterday.

jroweboy commented:
> Is CMakeDependentOption not a built-in module?

Looks like it should be. Removed.

> Specify that clang-format 6.0 is used in the contrib guide since different versions have different behaviours.

Good idea. Updated.

> Fairly sure this doesn't work on macOS (currently away from a mac for two days). If one wants to merge this before I check for contrib reasons, I don't mind, I'll check when I'm able.

Changed it to check for both `clang-format-6.0` and `clang-format` which isn't ideal, but is better than nothing

> Good timing. macbrew removed the clang-format@3.8 package yesterday.
👍

bunnei commented:
👍 ty lgtm
